### PR TITLE
improv(batch): improve errors

### LINKED ISSUE #1647: Feature request: improve error handling
### Use case

Currently some of the errors in the Batch Processing utility are either generic or hard to parse/read. We could improve on them considering that we encourage JSON-structured logging.

### Solution/User Experience

Currently the error thrown when the full batch fails processing looks like this:
```json
{
    "errorType": "Error",
    "errorMessage": "\nAll records failed processing. 1 individual errors logged separately below.\n,Failed to process record, 3\n",
    "msg": "All records failed processing. 1 individual errors logged separately below.",
    "childErrors": [
        {
            "errorType": "Error",
            "errorMessage": "Failed to process record, 3",
            "stack": [
                "Error: Failed to process record, 3",
                "    at BatchProcessor2.recordHandler [as handler] (/var/task/index.js:416:11)",
                "    at BatchProcessor2.processRecord (/var/task/index.js:250:31)",
                "    at BatchProcessor2.process (/var/task/index.js:131:38)",
                "    at processPartialResponse2 (/var/task/index.js:299:18)",
                "    at Runtime.handler (/var/task/index.js:421:50)",
                "    at Runtime.handleOnceNonStreaming (file:///var/runtime/index.mjs:1083:29)"
            ]
        }
    ],
    "stack": [
        "Error: ",
        "All records failed processing. 1 individual errors logged separately below.",
        ",Failed to process record, 3",
        "",
        "    at BatchProcessor2.clean (/var/task/index.js:179:17)",
        "    at BatchProcessor2.process (/var/task/index.js:133:14)",
        "    at processPartialResponse2 (/var/task/index.js:299:18)",
        "    at Runtime.handler (/var/task/index.js:421:50)",
        "    at Runtime.handleOnceNonStreaming (file:///var/runtime/index.mjs:1083:29)"
    ]
}
```

The error message could use some improvement and be rendered instead like:
```json
{
    "errorType": "FullBatchFailureError",
    "errorMessage": "All records failed processing. See individual errors below.",
    "name": "FullBatchFailureError",
    "recordErrors": [
        {
            "errorType": "Error",
            "errorMessage": "Failed to process record, 2",
            "stack": [
                "Error: Failed to process record, 2",
                "    at BatchProcessor2.recordHandler [as handler] (/var/task/index.js:421:11)",
                "    at BatchProcessor2.processRecord (/var/task/index.js:256:31)",
                "    at BatchProcessor2.process (/var/task/index.js:137:38)",
                "    at processPartialResponse2 (/var/task/index.js:304:18)",
                "    at Runtime.handler (/var/task/index.js:426:50)",
                "    at Runtime.handleOnceNonStreaming (file:///var/runtime/index.mjs:1083:29)"
            ]
        },
        {
            "errorType": "Error",
            "errorMessage": "Failed to process record, 3",
            "stack": [
                "Error: Failed to process record, 3",
                "    at BatchProcessor2.recordHandler [as handler] (/var/task/index.js:421:11)",
                "    at BatchProcessor2.processRecord (/var/task/index.js:256:31)",
                "    at BatchProcessor2.process (/var/task/index.js:137:38)",
                "    at processPartialResponse2 (/var/task/index.js:304:18)",
                "    at Runtime.handler (/var/task/index.js:426:50)",
                "    at Runtime.handleOnceNonStreaming (file:///var/runtime/index.mjs:1083:29)"
            ]
        },
        {
            "errorType": "Error",
            "errorMessage": "Failed to process record, 4",
            "stack": [
                "Error: Failed to process record, 4",
                "    at BatchProcessor2.recordHandler [as handler] (/var/task/index.js:421:11)",
                "    at BatchProcessor2.processRecord (/var/task/index.js:256:31)",
                "    at BatchProcessor2.process (/var/task/index.js:137:38)",
                "    at processPartialResponse2 (/var/task/index.js:304:18)",
                "    at Runtime.handler (/var/task/index.js:426:50)",
                "    at Runtime.handleOnceNonStreaming (file:///var/runtime/index.mjs:1083:29)"
            ]
        }
    ],
    "stack": [
        "FullBatchFailureError: All records failed processing. See individual errors below.",
        "    at BatchProcessor2.clean (/var/task/index.js:185:17)",
        "    at BatchProcessor2.process (/var/task/index.js:139:14)",
        "    at processPartialResponse2 (/var/task/index.js:304:18)",
        "    at Runtime.handler (/var/task/index.js:426:50)",
        "    at Runtime.handleOnceNonStreaming (file:///var/runtime/index.mjs:1083:29)"
    ]
}
```

Likewise, other errors thrown when the utility receives a wrong type, or short-circuits the processing of a FIFO SQS batch, can be improved either in their implementation or output.

### Alternative solutions

_No response_

### Acknowledgment

- [X] This feature request meets [Powertools for AWS Lambda (TypeScript) Tenets](https://docs.powertools.aws.dev/lambda-typescript/latest/#tenets)
- [ ] Should this be considered in other Powertools for AWS Lambda languages? i.e. [Python](https://github.com/aws-powertools/powertools-lambda-python/), [Java](https://github.com/aws-powertools/powertools-lambda-java/), and [.NET](https://github.com/aws-powertools/powertools-lambda-dotnet/)

### Future readers

Please react with 👍 and your use case to help us understand customer demand.

### PR DESCRIPTION
**Pull request by dreamorosi**
## Description of your changes

This PR improves the implementation of the errors thrown by the Batch Processing utility as well as some of the outputs by streamlining the log emitted or adding names to the errors and establishing a hierarchy.

Most of the errors were either using a generic `Error` class and/or expecting the caller to do the heavy-lifting of formatting the message. This PR introduces errors classes that are thrown in specific cases, which will help customers with `try/catch` logic.

Additionally, in most cases the errors are now performing the formatting or including the message/name that defines them without the caller having to specify it.

### Related issues, RFCs

**Issue number:** closes #1647

## Checklist

- [x] [My changes meet the tenets criteria](https://docs.powertools.aws.dev/lambda-typescript/#tenets)
- [x] I have performed a *self-review* of my own code
- [x] I have *commented* my code where necessary, particularly in areas that should be flagged with a TODO, or hard-to-understand areas
- [ ] I have made corresponding changes to the *documentation*
- [x] My changes generate *no new warnings*
- [ ] I have *added tests* that prove my change is effective and works
- [x] The PR title follows the [conventional commit semantics](https://github.com/aws-powertools/powertools-lambda-typescript/blob/main/.github/semantic.yml#L2)

### Breaking change checklist

***Is it a breaking change?:*** NO

- [ ] I have documented the migration process
- [ ] I have added, implemented necessary warnings (if it can live side by side)

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

### CODE REVIEW

File: packages/batch/src/errors.ts [hunk 1]
```diff
@@ -1,49 +1,58 @@
...
-class BaseBatchProcessingError extends Error {
-  public childErrors: Error[];
-
-  public msg: string;
-
-  public constructor(msg: string, childErrors: Error[]) {
-    super(msg);
-    this.msg = msg;
-    this.childErrors = childErrors;
+class BatchProcessingError extends Error {
```
Comment (erikayao93) [inline packages/batch/src/errors.ts:6]: Would we want the error thrown in the `BatchProcessor` [here](https://github.com/aws-powertools/powertools-lambda-typescript/blob/3ad0f5cffad7eb03fc2ac6deb78a29a1fad1a7fa/packages/batch/src/BatchProcessor.ts#L26C2-L26C2) to be as this `BatchProcessingError` instead of just `Error`, now that we have this?

File: packages/batch/tests/unit/BatchProcessor.test.ts
Comment (erikayao93) [inline packages/batch/tests/unit/BatchProcessor.test.ts:1]: Do we want a test or two to check the contents of errors? 
Same for the SQS FIFO processor class, we don't have anything that actually checks the error name, message, etc, of the error thrown.

**Review by erikayao93 — COMMENTED** [review]
I like the idea of the errors here, and I had some time to check in, so I thought I'd leave some comments! Feel free to let me know your thoughts.

File: packages/batch/src/errors.ts (hunk 1 above)
Comment (dreamorosi) [reply packages/batch/src/errors.ts:6]: Indeed, very good catch. Updated in the latest commit.

File: packages/batch/tests/unit/BatchProcessor.test.ts
Comment (dreamorosi) [reply packages/batch/tests/unit/BatchProcessor.test.ts:1]: I think it's a good point, I have updated the SQSFifo & other tests to at least test the error type.

For now we'll leave the Error contents out of the tests to avoid making them too rigid.

File: packages/batch/src/BatchProcessor.ts
Comment (erikayao93) [inline packages/batch/src/BatchProcessor.ts:1]: Do we want to update the error passed to the failure handler here (and in `AsyncBatchProcessor`) to be the `BatchProcessingError` as well?

And potentially add tests for the error type checking like with the `SqsFifoPartialProcessor` tests if we add the typing here?

Comment (dreamorosi) [reply packages/batch/src/BatchProcessor.ts:1]: I think we should leave `Error` there because that's the type expected by `failureHandler` and we want that method to stay generic so that customers [creating their own partial processor](https://docs.powertools.aws.dev/lambda/typescript/latest/utilities/batch/#create-your-own-partial-processor) can pass any type of error. Making the error more strict would force them to have to subclass our own error class.

**Review by erikayao93 — APPROVED** [review]
